week04_approx_rl/homework_pytorch_main.ipynb

```diff
@@ -1,1329 +1,1531 @@
+      "metadata": {
+        "id": "kVV72AB-soUH"
+      },
+      "source": [
+        "## About hyperparameters:\n",
+        "\n",
+        "The task has something in common with supervised learning: loss is optimized through the buffer (instead of Train dataset). But the distribution of states and actions in the buffer **is not stationary** and depends on the policy that generated it. It can even happen that the mean TD error across the buffer is very low but the performance is extremely poor (imagine the agent collecting data to the buffer always manages to avoid the ball).\n",
+        "\n",
+        "* Total timesteps and training time: It seems to be so huge, but actually it is normal for RL.\n",
+        "\n",
+        "* $\\epsilon$ decay shedule was taken from the original paper and is like traditional for epsilon-greedy policies. At the beginning of the training the agent's greedy policy is poor so many random actions should be taken.\n",
+        "\n",
+        "* Optimizer: In the original paper RMSProp was used (they did not have Adam in 2013) and it can work not worse than Adam. For us Adam was default and it worked.\n",
+        "\n",
+        "* lr: $10^{-3}$ would probably be too huge\n",
+        "\n",
+        "* batch size: This one can be very important: if it is too small the agent can fail to learn. Huge batch takes more time to process. If batch of size 8 can not be processed on the hardware you use take 2 (or even 4) batches of size 4, divide the loss on them by 2 (or 4) and make optimization step after both backward() calls in torch.\n",
+        "\n",
+        "* target network update frequency: has something in common with learning rate. Too frequent updates can lead to divergence. Too rare can lead to slow leraning. For millions of total timesteps thousands of inner steps seem ok. One iteration of target network updating is an iteration of the (this time approximate) $\\gamma$-compression that stands behind Q-learning. The more inner steps it makes the more accurate is the compression.\n",
+        "* max_grad_norm - just huge enough. In torch clip_grad_norm also evaluates the norm before clipping and it can be convenient for logging."
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "Plp8WC_esoUH"
+      },
+      "source": [
+        "### Video"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "id": "DdExc_AssoUH"
+      },
+      "outputs": [],
+      "source": [
+        "# record sessions\n",
+        "from gymnasium.wrappers import RecordVideo\n",
+        "\n",
+        "with RecordVideo(env=make_env(), video_folder='./videos',\n",
+        "                 episode_trigger = lambda episode_number: True) as env_monitor:\n",
+        "    sessions = [evaluate(env_monitor, agent, n_games=n_lives,\n",
+        "                         greedy=True) for _ in range(10)]\n",
+        "env.close()"
```

Comment:
Instead of calling `env.close()` explicitly, please use a `with` block instead:

```
with make_env() as env, RecordVideo(env=env, ...) as env_monitor:
    ...
```

Reply:
Fixed